### PR TITLE
Fix ambiguity of atomic accessor operator[].

### PR DESCRIPTION
**Pull request by fodinabor**
As reported in https://github.com/illuhad/hipSYCL/issues/380#issuecomment-803572968 [1], the `operator[]` is ambiguous, if using an atomic accessor.

### CODE REVIEW

**Comment by illuhad** [comment]
Actually we implement the interface exactly as described by the SYCL 2020 spec :)
https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_interface_for_buffer_command_accessors

It seems that there is now an oversight in the spec w.r.t. to the conditions in which to enable `operator[]`, presumably because `atomic` is completely deprecated...

File: tests/sycl/accessor.cpp [hunk 1]
```diff
@@ -161,6 +161,17 @@ BOOST_AUTO_TEST_CASE(accessor_api) {
...
+  queue.submit([&](s::handler& cgh) {
+    run_test([&](auto buf, auto... args) {
+      return buf.template get_access<s::access::mode::atomic>(cgh, args...);
+    });
+    // mostly compilation test
+    auto atomicAcc = buf_a.template get_access<s::access::mode::atomic>(cgh);
+    cgh.single_task<class accessor_api_atomic_device_accessors>([atomicAcc](){
+      atomicAcc[0].exchange(0);
+    });
+  });
```
Comment (illuhad) [inline tests/sycl/accessor.cpp:182]: Philosophical question: Do we actually need to/Should we test deprecated features?

File: include/hipSYCL/sycl/libkernel/accessor.hpp [hunk 2]
```diff
@@ -537,15 +537,17 @@ class accessor : public detail::accessor_base<std::remove_const_t<dataT>> {
...
   HIPSYCL_UNIVERSAL_TARGET
   reference operator[](id<dimensions> index) const
   {
     return (this->_ptr.get())[detail::linear_id<dimensions>::get(index, _buffer_range)];
   }
 
   template<int D = dimensions,
-            std::enable_if_t<(D == 1), bool> = true>
+            access::mode M = accessmode,
+            std::enable_if_t<(D == 1) && (M != access::mode::atomic), bool> = true>
```
Comment (illuhad) [inline include/hipSYCL/sycl/libkernel/accessor.hpp:566]: Wouldn't local accessor be affected as well?

Comment (fodinabor) [reply include/hipSYCL/sycl/libkernel/accessor.hpp:566]: only if they had atomic support in the first case (which they [deprecatedly should have as of SYCL 2020](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_accessor_specialization_with_targetlocal))

File: tests/sycl/accessor.cpp (hunk 1 above)
Comment (fodinabor) [reply tests/sycl/accessor.cpp:182]: I think as long as we did not remove them, we should test it, as we still provide it and the user will expect it to work as long as it's there.

**Review by fodinabor — COMMENTED** [review]
The spec seems to be undecided how to deal with this..
```cpp
/// note the missing accessMode check
  /* Available only when: (dimensions > 0) */
  reference operator[](id<dimensions> index) const;

/// note the missing accessMode check
  /* Available only when: (dimensions > 1) */
  __unspecified__ &operator[](size_t index) const;

/// note the accessMode check
  /* Available only when: (accessMode != access_mode::atomic && dimensions == 1) */
  reference operator[](size_t index) const;

  /* Deprecated
  Available only when: (accessMode == access_mode::atomic && dimensions ==  0) */
  operator cl::sycl::atomic<dataT, access::address_space::global_space> () const;

  /* Deprecated
  Available only when: (accessMode == access_mode::atomic && dimensions == 1) */
  cl::sycl::atomic<dataT, access::address_space::global_space> operator[](
    id<dimensions> index) const;
```

Just had another look.. seems we're missing atomic support for multi-dimensional accessors in `subscript_proxy` in general..

Thus.. do we plan on keeping atomics around? If so, I will add the missing parts for local and subscript proxy

**Comment by fodinabor** [comment]
@hipsycl-ci-bot test this now as well, please.

**Comment by illuhad** [comment]
@fodinabor Yes, I think when rewriting the spec we didn't really think about if/how things should play with deprecated stuff in this particular case...

I can get behind continuing to support for now what we already have, but not sure if it's a good idea to complicate `subscript_proxy` just for some deprecated feature...

EDIT: Okay, I see you already did the `subscript_proxy`. Nevermind then, so be it ;)

**Review by illuhad — APPROVED** [review]
And I was so happy that I could remove all those checks when I changed the interfaces to SYCL 2020 definitions... Can't wait for atomic accessors to go away for good :)

**Pull request merged**